spec.html

```diff
@@ -32455,23 +32456,37 @@ <h1>Runtime Semantics: RegExpInitialize ( _obj_, _pattern_, _flags_ )</h1>
-              1. Parse _pText_ using the grammars in <emu-xref href="#sec-patterns"></emu-xref>. The goal symbol for the parse is |Pattern[+U, +N]|. Throw a *SyntaxError* exception if _pText_ did not conform to the grammar, if any elements of _pText_ were not matched by the parse, or if any Early Error conditions exist.
-              1. Let _patternCharacters_ be a List whose elements are the code points of _pText_.
+              1. Let _patternText_ be the result of interpreting each of _P_'s 16-bit elements as a Unicode BMP code point. UTF-16 decoding is not applied to the elements.
+              1. Let _patternCharacters_ be a List whose elements are the code unit elements of _P_.
+            1. Let _parseResult_ be ParsePattern(_patternText_, _u_).
+            1. If _parseResult_ is a non-empty List of *SyntaxError* objects, throw a *SyntaxError* exception.
+            1. Assert: _parseResult_ is a Parse Node for |Pattern|.
             1. Set _obj_.[[OriginalSource]] to _P_.
             1. Set _obj_.[[OriginalFlags]] to _F_.
-            1. Set _obj_.[[RegExpMatcher]] to the Abstract Closure that evaluates the above parse by applying the semantics provided in <emu-xref href="#sec-pattern-semantics"></emu-xref> using _patternCharacters_ as the pattern's List of |SourceCharacter| values and _F_ as the flag parameters.
+            1. Set _obj_.[[RegExpMatcher]] to the Abstract Closure that evaluates _parseResult_ by applying the semantics provided in <emu-xref href="#sec-pattern-semantics"></emu-xref> using _patternCharacters_ as the pattern's List of |SourceCharacter| values and _F_ as the flag parameters.
             1. Perform ? Set(_obj_, *"lastIndex"*, 0, *true*).
             1. Return _obj_.
           </emu-alg>
         </emu-clause>
 
+        <emu-clause id="sec-parsepattern" aoid="ParsePattern">
+          <h1>Static Semantics: ParsePattern ( _patternText_, _u_ )</h1>
+          <p>The abstract operation ParsePattern takes arguments _patternText_ (a sequence of Unicode code points) and _u_ (a Boolean). It performs the following steps when called:</p>
+          <emu-alg>
+            1. If _u_ is *true*, then
+              1. Parse _patternText_ using the grammars in <emu-xref href="#sec-patterns"></emu-xref>. The goal symbol for the parse is |Pattern[+U, +N]|.
+            1. Else,
+              1. Parse _patternText_ using the grammars in <emu-xref href="#sec-patterns"></emu-xref>. The goal symbol for the parse is |Pattern[~U, ~N]|. If the result of parsing contains a |GroupName|, reparse with the goal symbol |Pattern[~U, +N]| and use this result instead.
+            1. If _patternText_ did not conform to the grammar, or any elements of _patternText_ were not matched by the parse, or any Early Error conditions exist, return a List of one or more *SyntaxError* objects representing the parsing errors and/or early errors.
+            1. Otherwise, return the Parse Node resulting from the parse.
```

Comment:
Can't we give this an alias?

Reply:
We *could*, but none of the original code does, and adding that wasn't necessary to the refactoring.

I'm thinking of a separate PR that introduces an abstract operation to parse a given text wrt a given goal symbol and analyze for early errors. That would concentrate most of the weirdness/wordiness around parsing in a single place. It would also make it easier to use an alias in places like the above.

Reply:
I look forward to such a PR.